File: src/koru/ide_adapters/registry.py
"""Registry of per-IDE adapters."""

from __future__ import annotations

from koru.ide_adapters.vscode_family import VSCodeFamilyAdapter

_ADAPTERS: dict[str, VSCodeFamilyAdapter] = {
    "cursor": VSCodeFamilyAdapter(
        ide_id="cursor",
        label="Cursor",
        requires_trusted_publisher=True,
    ),
    "vscode": VSCodeFamilyAdapter(
        ide_id="vscode",
        label="VS Code",
        requires_trusted_publisher=True,
    ),
    "vscodium": VSCodeFamilyAdapter(
        ide_id="vscodium",
        label="VSCodium",
        requires_trusted_publisher=True,
    ),
    "windsurf": VSCodeFamilyAdapter(
        ide_id="windsurf",
        label="Windsurf",
        requires_trusted_publisher=False,
    ),
    "antigravity": VSCodeFamilyAdapter(
        ide_id="antigravity",
        label="Antigravity",
        requires_trusted_publisher=False,
    ),
}


def get_adapter(ide: str) -> VSCodeFamilyAdapter | None:
    return _ADAPTERS.get(ide)


def supported_adapter_ids() -> tuple[str, ...]:
    return tuple(_ADAPTERS.keys())

File: plugins/koru-autopilot-cursor/src/ides/registry.ts
/**
 * Lightweight per-IDE strategy registry.
 *
 * Strategies self-register at import time. Callers must use `getStrategy`
 * (returns `undefined` for IDEs without a dedicated module) so the legacy
 * fall-through code in `probe-ladder.ts` / `extension.ts` keeps working
 * for IDEs that have not been extracted yet.
 */

import type { IdeStrategy } from "./ide-strategy";

const REGISTRY = new Map<string, IdeStrategy>();

export function registerStrategy(strategy: IdeStrategy, opts: { override?: boolean } = {}): void {
  const id = strategy.id;
  if (!id) {
    throw new Error("IdeStrategy.id must be a non-empty string");
  }
  if (!opts.override && REGISTRY.has(id)) {
    throw new Error(`IdeStrategy for ${id} already registered`);
  }
  REGISTRY.set(id, strategy);
}

export function getStrategy(id: string | undefined): IdeStrategy | undefined {
  if (!id) return undefined;
  return REGISTRY.get(id.toLowerCase());
}

export function allStrategies(): IdeStrategy[] {
  return [...REGISTRY.values()];
}

export function detectIdeViaStrategies(appName: string): string | undefined {
  for (const strat of REGISTRY.values()) {
    const id = strat.detectIde(appName);
    if (id) return id;
  }
  return undefined;
}

/**
 * Eager-load per-IDE strategy modules so they self-register before any
 * caller asks for them.
 *
 * This plugin ships only the Cursor strategy by design. Each other
 * IDE has its own standalone VSIX (``koru-autopilot-vscode``,
 * ``koru-autopilot-vscodium``, ``koru-autopilot-windsurf``,
 * ``koru-autopilot-antigravity``) so a regression here cannot leak
 * into another IDE's runtime.
 */
export function bootstrapStrategies(): void {
  // The require() form means TypeScript compiles to commonjs and each
  // strategy's top-level `registerStrategy(...)` call runs once.
  for (const mod of [
    "./cursor",
  ]) {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    require(mod);
  }
}

bootstrapStrategies();
